command/ca/revoke.go

```diff
@@ -0,0 +1,391 @@
+		if err := tok.UnsafeClaimsWithoutVerification(&claims); err != nil {
+			return nil, errors.Wrap(err, "error parsing flag '--token'")
+		}
+		if strings.ToLower(claims.Subject) != strings.ToLower(serial) {
+			return nil, errors.Errorf("token subject '%s' and serial number '%s' do not match", claims.Subject, serial)
+		}
+
+		// Prepare client for bootstrap or provisioning tokens
+		if len(claims.SHA) > 0 && len(claims.Audience) > 0 && strings.HasPrefix(strings.ToLower(claims.Audience[0]), "http") {
+			caURL = claims.Audience[0]
+			options = append(options, ca.WithRootSHA256(claims.SHA))
+			ui.PrintSelected("CA", caURL)
+			return ca.NewClient(caURL, options...)
+		}
+	}
+
+	if len(caURL) == 0 {
+		return nil, errs.RequiredFlag(ctx, "ca-url")
+	}
+	if len(rootFile) == 0 {
+		rootFile = pki.GetRootCAPath()
+		if _, err := os.Stat(rootFile); err != nil {
+			return nil, errs.RequiredFlag(ctx, "root")
+		}
+	}
+	options = append(options, ca.WithRootFile(rootFile))
+
+	ui.PrintSelected("CA", caURL)
+	return ca.NewClient(caURL, options...)
+}
+
+func (f *revokeFlow) GenerateToken(ctx *cli.Context, subject *string) (string, error) {
+	// For offline just generate the token
+	if f.offline {
+		return f.offlineCA.GenerateToken(ctx, revokeType, *subject, nil, time.Time{}, time.Time{})
+	}
+
+	// Use online CA to get the provisioners and generate the token
+	caURL := ctx.String("ca-url")
+	if len(caURL) == 0 {
+		return "", errs.RequiredUnlessFlag(ctx, "ca-url", "token")
+	}
+
+	root := ctx.String("root")
+	if len(root) == 0 {
+		root = pki.GetRootCAPath()
+		if _, err := os.Stat(root); err != nil {
+			return "", errs.RequiredUnlessFlag(ctx, "root", "token")
+		}
+	}
+
+	var err error
+	if *subject == "" {
+		*subject, err = ui.Prompt("What is the Serial Number of the certificate you would like to revoke? (`step certificate inspect foo.crt`)", ui.WithValidateNotEmpty())
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return newTokenFlow(ctx, revokeType, *subject, nil, caURL, root, time.Time{}, time.Time{})
+}
+
+func (f *revokeFlow) Revoke(ctx *cli.Context, serial, token string) error {
+	client, err := f.getClient(ctx, serial, token)
+	if err != nil {
+		return err
+	}
+
+	reason := ctx.String("reason")
+	// Convert the reasonCode flag to an OCSP revocation code.
+	reasonCode, err := ReasonCodeToNum(ctx.String("reasonCode"))
+	if err != nil {
+		return err
+	}
+
+	var tr http.RoundTripper
+
+	// If token is not provided then set up mTLS client with expected crt and key.
+	if len(token) == 0 {
+		crtFile, keyFile := ctx.String("crt"), ctx.String("key")
+
+		// If there is no token then we must be doing a Revoke over mTLS.
+		cert, err := tls.LoadX509KeyPair(crtFile, keyFile)
+		if err != nil {
+			return errors.Wrap(err, "error loading certificates")
+		}
+		if len(cert.Certificate) == 0 {
+			return errors.New("error loading certificate: certificate chain is empty")
+		}
+		rootCAs, err := x509util.ReadCertPool(ctx.String("root"))
```

Comment:
Sometimes root is loaded from the default location

Reply:
I don't understand what you are recommending / what you would like changed ??

Reply:
https://github.com/smallstep/cli/blob/master/command/ca/certificate.go#L286-L292